drivers/input/input_analog_axis.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT analog_axis
+
+#include <stdlib.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/input/input.h>
+#include <zephyr/input/input_analog_axis.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(analog_axis, CONFIG_INPUT_LOG_LEVEL);
+
+#define ANALOG_AXIS_SETTINGS_PATH_MAX 32
+
+struct analog_axis_channel_config {
+	struct adc_dt_spec adc;
+	int16_t out_min;
+	int16_t out_max;
+	uint16_t axis;
+	bool invert;
+};
+
+struct analog_axis_channel_data {
+	int last_out;
+};
+
+struct analog_axis_config {
+	uint32_t poll_period_ms;
+	const struct analog_axis_channel_config *channel_cfg;
+	struct analog_axis_channel_data *channel_data;
+	struct analog_axis_calibration *calibration;
+	const uint8_t num_channels;
+};
+
+struct analog_axis_data {
+	struct k_mutex cal_lock;
+	analog_axis_raw_data_t raw_data_cb;
+	struct k_timer timer;
+	struct k_thread thread;
+
+	K_KERNEL_STACK_MEMBER(thread_stack,
+			      CONFIG_INPUT_ANALOG_AXIS_THREAD_STACK_SIZE);
+};
+
+static void analog_axis_calibration_log(const struct device *dev)
+{
+	const struct analog_axis_config *cfg = dev->config;
+	int i;
+
+	for (i = 0; i < cfg->num_channels; i++) {
+		struct analog_axis_calibration *cal = &cfg->calibration[i];
+
+		LOG_INF("%s: ch: %d min: %d max: %d deadzone: %d",
+			dev->name, i, cal->in_min, cal->in_max, cal->out_deadzone);
+	}
+}
+
+#ifdef CONFIG_SETTINGS
```

Comment:
I feel like making drivers dependent on settings is wrong, driver should just process data and maybe provide an API to feed calibration values (read from settings or wherever else).

Reply:
Yeah good point, it's a very convenient feature to have though and quite some boilerplate to implement though. What if I split it out in its own file? Maybe I can put it out of driver into `subsys/input/analog_axis_settings.c`?

Reply:
ok I've split the settings out in its own file, backed by a separate Kconfig option and only using the "public" driver API, left it in drivers though, lmk what you think.